adr/ADR-42.md

```diff
@@ -0,0 +1,168 @@
+
+Some examples:
+
+ * A single client should receive all messages from a consumer and if it fails another should take over
+ * Groups of related clients should be organised such that certain clients only receive messages when various conditions are met
+ * Groups of clients should access a Consumer but some should have higher priority than others for receiving messages
+ * A consumer should deliver messages to clients in a specific grouping such that related messages all go to the same client group
+
+The proposed feature here address some of these needs while paving the way for future needs to be addressed within this
+framework and under this umbrella feature called `Pull Consumer Groups`.
+
+The current focus is around providing building blocks that can be used to solve higher order problems client-side.
+
+Related prior work:
+
+ * Proposed [Consumer Groups](https://github.com/nats-io/nats-architecture-and-design/pull/36)
+ * Proposed [Partitioned consumer groups and exclusive consumer](https://github.com/nats-io/nats-architecture-and-design/pull/263)
+ * Proposed [Consumer Owner ID](https://github.com/nats-io/nats-server/pull/5157)
+
+## General Overview
+
+We introduce 2 settings on `ConsumerConfig` that activates these related features:
+
+```go
+{
+	PriorityGroups: ["<group>", ...],
+	PriorityPolicy: "<policy>"
+}
+```
+
+The presence of the `PriorityGroups` setting activates the set of features documented here with a list of names supported
+behaviours configured using the `PriorityPolicy`.
+
+Technically in some of the proposed policies the `PriorityGroups` have no real meaning today, but we keep it for consistency 
+and allow for future features to be added that would be per-group without then requiring all clients to be updated. 
+Future message grouping features would require groups to be listed here.
+
+In the initial implementation we should limit `PriorityGroups` to one per consumer only and error should one be made with 
+multiple groups. In future iterations multiple groups will be supported along with dynamic partitioning of stream
+data.
+
+This is only supported on Pull Consumers, configuring this on a Push consumer must raise an error.
+
+## Priority Policies
+
+### `overflow` policy
+
+Users want certain clients to only get messages when certain criteria are met.
+
+Imagine jobs are best processed locally in `us-east-1` but at times there might be so many jobs in that region that 
+`us-west-1` region handling some overflow, while less optimal in terms of transit costs and latency, would be desirable 
+to ensure serving client needs as soon as possible.
+
+The `overflow` policy enables Pull requests to be served only if criteria like `num_pending` and `num_ack_pending` are
+above a certain limit, otherwise those Pull requests will sit idle in the same way that they would if no messages were
+available (receiving heartbeats etc).
+
+```go
+{
+	PriorityGroups: ["jobs"],
+	PriorityPolicy: "overflow",
```

Comment:
This should also be an option on regular consumers

Reply:
I thought about that, but I really dont think we know all the implications - like for example just on the other policy we'd have to then error if these are added.

Seems like a can of worms of crazy matrics of features.  So I am really keen to move to a more clear intent when it comes to these things.  You want this behavior you gotta say so rather than depend on a 5 boolean if statement :)